## Importing

In [1]:
#libraries
import pandas as pd
import numpy as np
import os


In [4]:
#establishing file path short cut
path = r'C:\Users\evras\Desktop\Data Analysis\Immersion\A4 Python\Instacart Basket Analysis\02 Data'
#checking
path

'C:\\Users\\evras\\Desktop\\Data Analysis\\Immersion\\A4 Python\\Instacart Basket Analysis\\02 Data'

In [23]:
#bringing in the files
#Orders
df_ords = pd.read_csv(os.path.join(path,'Preped Data','orders_wrangled.csv'),index_col=False)

In [24]:
#checking file
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0           0   2539329        1             1                  2   
1           1   2398795        1             2                  3   
2           2    473747        1             3                  3   
3           3   2254736        1             4                  4   
4           4    431534        1             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [25]:
#looks like the index line stuck around? Need to justt snip that
df_ords = df_ords.drop(columns=['Unnamed: 0'])

In [26]:
#checking to make suer it is gone
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [16]:
#Prodcuts
df_prods = pd.read_csv(os.path.join(path,'Preped Data','products_clean.csv'),index_col=False)

In [12]:
#checking
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [29]:
#pesky index...
df_prods = df_prods.drop(columns=['Unnamed: 0'])

In [30]:
#double checky check
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

### #1

##### Did this during the lesson but I deleted the script stuff so, I just loaded in the cleaned/verified version of products up top.

### #2

In [31]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

##### $99,999 is an awful lost of money for some groceries...

### #3

In [33]:
#checking for mixed data types
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_ords[weird])>0:
        print(col)

In [34]:
#well it didn't print anything, so that's good. but I just want to check...
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

### #4

##### looks good, If there had been a column that was weird I would have used
df_ords['col_name')=df_ords['col_name'].astype(str or int64)

### #5

In [35]:
#time to look for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### #6

In [38]:
#206,209 seems like a lot, but it's just first orders so there isn't anything to be alarmed about
#I think I'd like to replace them with 0 honestly.
#removing that many lines would be WAY to impactful on the set. So let's give it a go...
df_ords=df_ords['days_since_prior_order'].fillna(0,inplace=True)

In [39]:
#time to see if it worked
df_ords.isnull().sum()

AttributeError: 'NoneType' object has no attribute 'isnull'

In [56]:
#HAHA WOOPS I wen't and ruined the whole thing, thats ok it's an easy fix this time. Hilarious.
#I looked it up and I should have specified the column on the left half too, and didnt need the inplace bit.
#thats ok, this how we learn. Sort of funny though. Suppose this is why they suggest copies haha.

In [50]:
#bringing back orders
df_ords = pd.read_csv(os.path.join(path,'Preped Data','orders_wrangled.csv'),index_col=False)
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [59]:
#out of how many I wonder...
df_ords.shape

(3421083, 7)

In [51]:
#ok, doing it the right way this time. Hopefully. Who knows.
df_ords['days_since_prior_order']=df_ords['days_since_prior_order'].fillna(0)

In [52]:
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
order_day_of_week         0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [58]:
#boom baby, nice.

##### I sort of mentioned it already but the reason I went with this route was because of two things. Firstly, when I was looking at the DF earlier durring the lesson I noticed the NaNs and realized it was just that the user hand't made a purchase before, so naturally there wouldn't be a value there. So the correct answer actually IS 0, even if it ends up being like 2/3 of the column. Secondly if I were to impute the value via mean or median it would skew the results of analysis dramatically, giving the illusion that people order with more or less frequency than they do. A vitally import piece of data given the set. Finally I didn't want to just remove the rows because it would mean an analysis of literally any other part ofthe day would be prfoundly off. We're talking 206209 rows, which is like 6%? That was too much to just get rid of when the best solution is just to turn it into an accurate entry. There is the posibility that some are not just first time purcheses, but I dobut it given that the data is likely collected by the app itself.
##### I say all this like I activly say here and pondered that out but, honestly it just clicked. I might be getting used to this jazz haha

### #7

In [60]:
#ok, time to look for the duplicates now
df_ords_dups = df_ords[df_ords.duplicated()]

In [61]:
#check that one out now
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [62]:
#well that was almost suspiciously easy.

##### well there were no dups, so nothing to do there. I would have used the following line though to solve it, for what its worth
##### df_ords_nodups = df_ords.drop_duplicates()

### #9

In [63]:
#last but not least, time to extract these puppies. Well, Ords anyway because I didn't do anything to
#prods. That one is still all shiney from the actual lesson.
df_ords.to_csv(os.path.join(path,'Preped Data','orders_shiney.csv'))